c/src/space.rs

```diff
@@ -1,12 +1,376 @@
 use hyperon::space::grounding::*;
-use hyperon::space::Space;
+use hyperon::space::*;
+use hyperon::atom::*;
+use hyperon::matcher::*;
 
 use crate::atom::*;
 use crate::util::*;
 
 use std::os::raw::*;
+use core::cell::RefCell;
+use std::rc::Rc;
 
-// GroundingSpace
+//-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-
+// Space & SpaceMut trait interface wrapper
+//-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-=-+-
+
+//QUESTION FOR VITALY: This C interface could be simplified a lot by making the CSpace object
+// automatically handles calls to the observer instead of exporting the mechanism to register
+// observers.  However, my reasoning was that there must be a good reason the observer
+// interaction is part of the Space trait, and not part of a wrapper around a Space, and that
+// was the reason  I carried the observer mechanism forward into this C API.
+//
+//However, if there isn't a situation where a given space needs to manipulate the events seen
+// by the observer (hiding some events?, creating new events?) then we can simplify the API as
+// well as eliminate a potential source of mistakes by storing observers and handling observer
+// interactions automatically for CSpaces.
```

Comment:
I see.  So we need a way for the space to notify observers manually, no matter what.

I was wondering if it makes sense for the CSpace to automatically notify observers when the add, remove, and replace calls are made, rather than relying on the C code to do it.  It's possible this is more trouble than its worth and might force the space implementor into an unnatural design to avoid double-notify bugs.

I'll leave it the way it is unless you disagree.

Reply:
> I was wondering if it makes sense for the CSpace to automatically notify observers when the add, remove, and replace calls are made, rather than relying on the C code to do it.

The point is C space can have rather a complex structure which may not always updated through calls to `add`, `modify` and `remove` as I said. But I also agree implementing observers even in Rust is cumbersome. I think we can think in two directions:
- split the work between PR's: make the simplest implementation possible in this PR and work on the full implementation in next PR;
- eliminate observers in code for now and postpone this until the next use-case.

Reply:
I think the direction reflected in https://github.com/trueagi-io/hyperon-experimental/issues/287 makes sense.